source-academy/frontend
# SICP JS: Rework program snippets

### Description



This PR is dependent on source-academy/sicp#545.

1. Fixes #1830.
2. Reworked how snippet dependencies are hidden. Dependencies are now hidden by code folding instead. This solves the environment model not working with dependencies bug and the bug where some program snippets cannot be run when dependencies are hidden because the prepend was evaluated in a different scope to the main body.

<img width="288" alt="Screenshot 2021-06-26 at 8 38 07 PM" src="https://user-images.githubusercontent.com/60355570/123513212-6ca7d200-d6be-11eb-9026-2968ac048703.png">

In the image above, taken from the third program snippet in section 3.2.3, lines 2 to 15 are hidden. A user can click on the yellow button to reveal dependencies.

<img width="496" alt="Screenshot 2021-06-26 at 8 39 09 PM" src="https://user-images.githubusercontent.com/60355570/123513235-919c4500-d6be-11eb-9aa0-b0f74a51962c.png">

With this solution the only way to hide dependencies again is to close and reopen the snippet.

### Type of change


### How to test



1. Ensure correct lines are highlighted when using env visualiser with and without code folding.
2. Ensure snippets with dependencies are hidden correctly and can be revealed by pressing the yellow button. The env viz should also work correctly when dependencies are hidden.

### Checklist
